simulator/infra/cache/cache_tag_array.h

```diff
@@ -110,7 +111,46 @@ class CacheTagArray : public CacheTagArraySize
         // hash tabe to lookup tags in O(1)
         std::vector<google::dense_hash_map<Addr, Way>> lookup_helper;
         const uint32 impossible_key = INT32_MAX;
-        ReplacementModule replacement_module; // LRU algorithm module
+        std::unique_ptr<ReplacementModule> replacement_module = nullptr;
+};
+
+class AlwaysHitCacheTagArray : public CacheTagArray
```

Comment:
The structural problem is that you inherit AlwaysHitCacheTagArray from CacheTagArray, but not from the common ancestor. The hierarchy may be:
TagArray ← CacheTagArray <= Replacement Module + Tag array
TagArray ← InfiniteTagArray <= dense_hash_map as a tag array instead of vector
TagArray ← AlwaysHitTagArray <= always returns true

Reply:
I dicided to inherit AlwaysHitCacheTagArray from CacheTagArray, because otherwise it would be big amount of copy-paste code, since AlwaysHitCacheTagArray e.g. ovverides only one of the four methods. I don't understand how the structure you've suggested me could fix the problem with `BP::directions` and `BP::targets` vectors

Reply:
Always-hit cache cannot provide the data, so we cannot use it for BP, but we can use it for instruction cache and data cache where we model only tags.

Reply:
So then everithing should be working right. In wich modules do I test infinite and always hit models?

Reply:
Do you mean unit test or just common sense test?

Reply:
I mean unit tests

Reply:
infra/cache/t